### PR TITLE
fix(server): surface MAX_TOOL_ROUNDS cap to model + user (#4063)

### PR DESCRIPTION
**Pull request by blamechris**
## Summary

When the BYOK agent loop hit the safety cap (25 rounds), it just `break`-ed silently. The dashboard saw a `tool_use` stop_reason on the last round and then nothing — looked like a hang.

Now, on cap-hit:

- **Non-fatal error event** — `{ type: 'error', code: 'MAX_TOOL_ROUNDS_REACHED', fatal: false, message: 'Tool round cap reached (25). Asked the model to summarise…' }` so the dashboard can render a warning banner without killing the session.
- **Synthetic user turn** — `"Tool budget exhausted (25 rounds). Please summarise what you accomplished so far. Do not call any more tools."` appended to history.
- **Summary stream** — one final `messages.stream({ messages, /* tools omitted */ })` so the model produces a text-only closing summary that the dashboard streams as deltas.
- **Accounting** — the summary round's `usage` + `cost` fold into the same `turnUsage` / `turnCost`, so the existing `result` event is accurate end-to-end.
- **Init-failure fallback** — if the summary stream's `messages.stream()` throws synchronously, the synthetic user message is popped off `_history` to avoid orphaning it, and the outer loop exits cleanly.

## Test plan

- [x] `breaks out of the loop after MAX_TOOL_ROUNDS` updated — asserts **26** calls (25 tool rounds + 1 summary)
- [x] `emits a non-fatal MAX_TOOL_ROUNDS_REACHED error when the cap fires` — pins the new error event shape + non-fatal flag + session-still-usable
- [x] `streams summary text from the post-cap round so the user sees what was accomplished` — pins the streamed delta + final `stopReason: 'end_turn'`
- [x] Full byok-session suite: 32/32 pass
- [x] Sibling agent-event-translator + tool-executor suites unaffected

Closes #4063.

### CODE REVIEW

**Review by blamechris — COMMENTED** [review]
## Verdict: APPROVE with follow-ups (non-blocking)

Solid PR — directly addresses #4063, CI green (13/13), tests pass (32/32 byok-session suite), no protocol-shape breakage. The cap-hit no longer looks like a hang. Code style matches server conventions (ES modules, no semicolons, single quotes).

### Counts
- Critical issues: 0
- Architectural concerns: 1
- Test gaps: 1
- UX/dashboard follow-up: 1
- Code-quality nits: 1

### Findings

#### 1. Same-role consecutive `user` turns (architectural — medium)

The cap handler appends a NEW `user` turn (synthetic instruction at `byok-session.js:460-466`) immediately after the tool_result `user` turn pushed at line 432. History layout becomes:

```
... assistant(tool_use) → user(tool_results) → user(synthetic) → assistant(summary)
```

The PR comment at line 458 says *"Per Anthropic API, two consecutive user turns are concatenated server-side; the synthetic message is part of the same logical turn."* — that's not actually documented API behavior. The API tolerates same-role consecutives but doesn't documentably concatenate them.

This is exactly the invariant `#4109` / `#4118` rollback paths were added to PROTECT. Comment at byok-session.js:295 says *"soft-breaking the alternation invariant the API may tighten on in future."*

**Cleaner alternative**: embed the synthetic instruction as an additional `text` content block on the EXISTING tool_result user turn:

```js
const summaryInstruction = { type: 'text', text: 'Tool budget exhausted (...). Please summarise...' }
this._history.push({ role: 'user', content: [...toolResults, summaryInstruction] })
```

That keeps strict alternation and removes reliance on undocumented behavior.

Tracked in #4146.

#### 2. Missing test coverage for cap-handler error paths (low)

Two new edge cases aren't pinned by tests:

- **Summary stream-init failure** (line 487 `this._history.length -= 1`) — no test forces `messages.stream()` to throw synchronously on the summary call.
- **Abort during summary for-await** — propagates to outer catch, which truncates and emits ABORT. The user sees TWO error events (`MAX_TOOL_ROUNDS_REACHED` + `ABORT`). Worth a test to pin the behavior, or a decision to suppress the duplicate.

Tracked in #4147.

#### 3. Dashboard does not act on `fatal: false` yet (UX — low)

`fatal: false` is wire-protocol-correct but the dashboard's generic error handler at `packages/dashboard/src/store/message-handler.ts:2886` calls `addServerError(surfaced)` regardless. Today `MAX_TOOL_ROUNDS_REACHED` renders identically to a real `STREAM_ERROR` (red toast / banner). The user gets the message but not the soft visual distinction the PR is setting up for.

Not blocking — the wire shape is correct; the UI branch can ship in a follow-up.

Tracked in #4148.

#### 4. Inline block could be extracted (nit)

The ~40-line summary-stream block at lines 468-516 duplicates the per-round model/maxTokens/system/messages signature and the usage accumulation pattern. A small private helper (`_runFinalRound({ messageId, systemPrompt, turnUsage, ... })`) would deduplicate and shrink the agent loop. Not blocking.

### Correctness — verified

- ✓ `pricing` is still in scope at the summary accumulation (declared at line 281, before the loop)
- ✓ `historyLengthBeforeSend` snapshot covers the synthetic user push — if anything throws inside the for-await or finalMessage, the outer catch (#4118) rolls back the entire turn cleanly
- ✓ Synthetic instruction text is fixed — no user input interpolation, no PII / secrets
- ✓ Stream-init failure rollback is correct (`length -= 1` undoes the synthetic push)
- ✓ Test rebase from `<= 25` to `=== 26` is the right number (25 tool rounds + 1 summary)

### Re: `stop_reason` if model misbehaves

The PR concern about "what if the summary returns tool_use" is moot in practice — the summary stream omits the `tools` parameter entirely, so the model can't emit tool_use blocks. Worth a sanity comment to make that explicit. The mock test could (and should) be more defensive here — see #4147.

---

Approving. Recommend landing as-is and addressing #4146 / #4147 / #4148 as follow-ups.

File: packages/server/src/byok-session.js [hunk 1]
```diff
@@ -434,8 +434,85 @@ export class ClaudeByokSession extends BaseSession {
...
+              case 'message_delta':
+                if (t.stopReason) lastStopReason = t.stopReason
+                break
+              default:
+                break
+            }
+          }
+
+          const summaryFinal = await summaryStream.finalMessage()
+          lastStopReason = summaryFinal.stop_reason
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/src/byok-session.js:521]: **FIX**

Fixed in `5d3808b26`

**Change:** Documented (and tested) that async failures inside the summary stream's for-await or finalMessage propagate to the existing outer try/catch — which already truncates `_history` to `historyLengthBeforeSend`, rolling back the entire cap-hit turn (consistent with #4109/#4118 patterns). New test `rolls back the entire turn when the summary stream rejects async` asserts this: history rolls back AND the user sees both MAX_TOOL_ROUNDS_REACHED + STREAM_ERROR error events.

The for-await + finalMessage() are now wrapped under the same outer try, with a comment noting the propagation path:

```diff
+          // Async failures inside the for-await or finalMessage() rethrow
+          // and propagate to the outer catch, which truncates _history to
+          // historyLengthBeforeSend — rolling back the entire turn
+          // (consistent with #4109/#4118).
```

File: packages/server/src/byok-session.js [hunk 2]
```diff
@@ -434,8 +434,85 @@ export class ClaudeByokSession extends BaseSession {
...
+              { signal: this._abortController.signal },
+            )
+          } catch (err) {
+            // Stream-init threw on the summary round. Bail without breaking
+            // alternation — the user instruction we just pushed is the
+            // last turn; truncating is safe because the next sendMessage
+            // would start with a new user prompt anyway.
+            this._history.length -= 1
+            log.warn(`MAX_TOOL_ROUNDS summary stream-init failed: ${err?.message}`)
+            break
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/src/byok-session.js:496]: **FIX**

Fixed in `5d3808b26`

**Change:** Two changes — (a) the synthetic message is now embedded as a text content block on the existing tool_result user turn instead of pushing a new turn (so the alternation invariant is preserved); (b) sync stream-init rollback uses `pop()` + identity check, much clearer than the prior `length -= 1`:

```diff
-            this._history.length -= 1
+            const popped = lastTurn.content.pop()
+            if (popped !== summaryInstruction) {
+              log.warn(`MAX_TOOL_ROUNDS rollback: popped content block was not the summary instruction`)
+            }
```

Pinned with `pops the synthetic instruction on summary stream-init failure` — asserts the user turn returns to pure tool_result blocks after failure.
